skorch/net.py

```diff
@@ -1778,29 +1977,44 @@ def _register_attribute(
 
         Parameters
         ----------
+        name : str
+          Name of the attribute.
+
+        attr : torch.nn.Module or torch.optim.Optimizer
+          The attribute itself.
+
         prefixes : bool (default=True)
           Whether to add to prefixes_.
 
         cuda_dependent_attributes : bool (default=True)
           Whether to add to cuda_dependent_attributes_.
 
         """
-        # copy the lists to avoid mutation
+        name = name.rstrip('_')  # module_ -> module
+
+        # Always copy the collections to avoid mutation
         if prefixes:
             self.prefixes_ = self.prefixes_[:] + [name]
 
         if cuda_dependent_attributes:
             self.cuda_dependent_attributes_ = (
                 self.cuda_dependent_attributes_[:] + [name + '_'])
 
+        if self.init_context_ == 'module':
```

Comment:
From what I can tell we need the context manager because of how we are using `__setattr__` to update the state of `modules_` and friends. Is this correct?

Reply:
Yes, exactly. This is my proposal how to "know" what kind of attribute we're dealing with, the old one being to infer from the attribute name, which I find inferior. A small disadvantage is that the context is not triggered if a user were to call `initialize_module` et al directly, though I don't believe there is a need for that.